### PR TITLE
Input sanitization

### LINKED ISSUE #46: XML type safety
What is the semantics of XString here? Is it supposed to be a piece of XML (in which case it should be a newtype and not an alias over `String`)? Or is it supposed to be any, potentially unescaped, string?

https://github.com/wireapp/hsaml2/blob/master/SAML2/Core/Assertions.hs#L49-L53

### PR DESCRIPTION
**Pull request by fisx**
Fixes #46

Texts in application types are `XmlText` (an opaque type that handles XML serialization robustly).  Start reading this PR from module `SAML2.WebSSO.Types`.

### CODE REVIEW

**Comment by fisx** [comment]
(still doesn't compile, but the `Types` module does, and it should be clear where this is going.)

File: src/SAML2/WebSSO/SP.hs [hunk 1]
```diff
@@ -297,15 +297,17 @@ checkAssertions missuer (toList -> assertions) uref@(UserRef issuer _) = do
   when (null statements) $
     deny DeniedNoStatements  -- (not sure this is even possible?)
 
-  when (null . catMaybes $ (^? _AuthnStatement) <$> statements) $
+  unless (any isAuthnStatement statements) $
```
Comment (fisx) [inline src/SAML2/WebSSO/SP.hs]: this change looked necessary in some intermediate state of things, now it *could* be rolled back.  i kinda like the new version, though.

File: src/SAML2/WebSSO/Types.hs [hunk 2]
```diff
@@ -1,7 +1,140 @@
...
+  , Locality(..)
+  , localityAddress
+  , localityDNSName
+  , normalizeAssertion
+  , idPIdToST
+  , assEndOfLife
+  , rspInResponseTo
+  , getUserRef
+  , nelConcat
+  , (<$$>)
```
Comment (fisx) [inline src/SAML2/WebSSO/Types.hs]: I have considered moving `XmlText` to a new module so I can skip the explicit exports in this module, but decided that explicit exports are nice, even if they are a bit long.  (Nothing I feel too strongly about, though.)

Comment (neongreen) [reply src/SAML2/WebSSO/Types.hs]: I like them and would even enforce them on the project-wide level. Having sections would also be nice, though.

File: src/SAML2/WebSSO/Types.hs [hunk 3]
```diff
@@ -25,7 +157,48 @@ import URI.ByteString  -- FUTUREWORK: should saml2-web-sso also use the URI from
...
 import qualified Data.X509 as X509
+import qualified Foundation.Network.IPv4 as IPv4
+import qualified Foundation.Network.IPv6 as IPv6
+import qualified Foundation.Parser as IP
+import qualified Network.DNS.Utils as DNS
 import qualified Servant
+import qualified Text.Email.Validate as Email
+
+
+-- | Text that needs to be escaped when rendered into XML.  See 'mkXmlText', 'mkXmlText'.
```
Comment (neongreen) [inline src/SAML2/WebSSO/Types.hs]: Might also be worth mentioning that `Text` is not allowed to be anywhere in types anymore.

Comment (neongreen) [inline src/SAML2/WebSSO/Types.hs]:
```suggestion
-- | Text that needs to be escaped when rendered into XML.  See 'mkXmlText', 'escapeXmlText'.
```

File: src/SAML2/WebSSO/Types.hs [hunk 4]
```diff
@@ -25,7 +157,48 @@ import URI.ByteString  -- FUTUREWORK: should saml2-web-sso also use the URI from
...
+newtype XmlText = XmlText { fromXmlText :: ST }
+  deriving (Eq, Ord, Show, Generic)
+
+mkXmlText :: ST -> XmlText
```
Comment (neongreen) [inline src/SAML2/WebSSO/Types.hs]: This means that `&lt;` is not allowed anywhere in strings that we get from outside, right?

It actually seems kinda weird and I don't think it makes anything better. I would just do `mkXmlText = XmlText` and whenever an issue pops up, it's a bug that needs to be fixed. Unescaping-and-escaping just hides it.

Comment (fisx) [reply src/SAML2/WebSSO/Types.hs]: Yes, I got that wrong, thanks!  The xml parsers definitely should give us unescaped content, and if not, we should fix it there.

Tasks:

- [ ] test case for NameID with formats unspecified, email, entity (aka uri), with escaping corner cases.
- [ ] factor `unescapeXmlText` out of `mkXmlText`
- [ ] `mkXmlText = XmlText`, as you suggest.

File: src/SAML2/WebSSO/SP.hs (hunk 1 above)
Comment (ChrisPenner) [reply src/SAML2/WebSSO/SP.hs]: I agree that this one is better;

For future reference the old version can be expressed a little clearer as 

```haskell
unless (has (traversed . _AuthnStatement) statements)
```

File: test/Test/SAML2/WebSSO/XMLSpec.hs [hunk 5]
```diff
@@ -54,15 +54,15 @@ spec = describe "XML Sanitization" $ do
         `shouldBe` (emailNameID "somebody@example.org" :: Either String NameID)
 
     it "should fail to decode an invalid email" $ do
-      decodeElem @NameID (xmlWithName emailFormat "&lt;somebody@example.org&gt;")  
+      decodeElem @NameID (xmlWithName emailFormat "&lt;somebody@example.org&gt;")
```
Comment (fisx) [inline test/Test/SAML2/WebSSO/XMLSpec.hs]: @ChrisPenner if you like, you can teach your vim to do that for us.  If not, don't bother, I have "write an integration test for this" somewhere on a TODO list.  ;-)

**Comment by fisx** [comment]
@ChrisPenner all yours again!

File: test/Test/SAML2/WebSSO/XMLSpec.hs (hunk 5 above)
Comment (ChrisPenner) [reply test/Test/SAML2/WebSSO/XMLSpec.hs]: Sorry, what are you referring to? trailing whitespace issues or something else?

Comment (fisx) [reply test/Test/SAML2/WebSSO/XMLSpec.hs]: > Sorry, what are you referring to? trailing whitespace issues or something else?

sorry, yes.  trailing whitespace issues.
